### ISSUE TITLE
ValueError: Need atleast one array to stack

### ISSUE THREAD

**Issue by gopalkumr** [tudelft-iv/view-of-delft-dataset#41]
I'm trying to run 4_evaluation.ipynb to evalute my model but it's giving Value error. Please help

<img width="1080" alt="image" src="https://user-images.githubusercontent.com/91552766/229763544-ddccbd6d-b0dc-48cc-bc56-4f3e97d9cbf7.png">

> > ValueError                                Traceback (most recent call last)
Cell In[4], line 9
      6 evaluation = Evaluation(test_annotation_file=os.path.join('[E:/vod/view_of_delft_PUBLIC](file:///E:/vod/view_of_delft_PUBLIC)', 'label'))
      8 # Using the evaluate method, the model can be evaluated on the detection labels.
----> 9 results = evaluation.evaluate(
     10     #result_path=os.path.join('example_set', 'detection'),
     11     result_path=os.path.join('[E:/vod/view_of_delft_PUBLIC](file:///E:/vod/view_of_delft_PUBLIC)', 'detection'),
     12     current_class=[0, 1, 2])
     14 print("Results: \n"
     15       f"Entire annotated area: \n"
     16       f"Car: {results['entire_area']['Car_3d_all']} \n"
   (...)
     24       f"mAP: {(results['roi']['Car_3d_all'] + results['roi']['Pedestrian_3d_all'] + results['roi']['Cyclist_3d_all']) / 3} \n"
     25       )

File [e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\evaluate.py:44](file:///E:/github/view-of-delft-dataset/view-of-delft-dataset/vod/evaluation/evaluate.py:44), in Evaluation.evaluate(self, result_path, current_class, score_thresh)
     41 gt_annotations = kitti.get_label_annotations(self.test_annotation_file, val_image_ids)
     43 evaluation_result = {}
---> 44 evaluation_result.update(get_official_eval_result(gt_annotations, dt_annotations, current_class, custom_method=0))
     45 evaluation_result.update(get_official_eval_result(gt_annotations, dt_annotations, current_class, custom_method=3))
     47 return evaluation_result

File [e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\kitti_official_evaluate.py:744](file:///E:/github/view-of-delft-dataset/view-of-delft-dataset/vod/evaluation/kitti_official_evaluate.py:744), in get_official_eval_result(gt_annotations, dt_annotations, current_classes, pr_detail_dict, custom_method)
    741             compute_aos = True
    742         break
--> 744 mAPbbox, mAPbev, mAP3d, mAPaos, mAPbbox_R40, mAPbev_R40, mAP3d_R40, mAPaos_R40 = do_eval(
    745     gt_annotations, dt_annotations, current_classes, min_overlaps, compute_aos, pr_detail_dict=pr_detail_dict,
    746     custom_method=custom_method)
    748 ret_dict = {}
    749 for j, curcls in enumerate(current_classes):
    750     # mAP threshold array: [num_min_overlap, metric, class]
    751     # mAP result: [num_class, num_diff, num_min_overlap]

File [e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\kitti_official_evaluate.py:650](file:///E:/github/view-of-delft-dataset/view-of-delft-dataset/vod/evaluation/kitti_official_evaluate.py:650), in do_eval(gt_annotations, dt_annotations, current_classes, min_overlaps, compute_aos, pr_detail_dict, custom_method)
    647 if custom_method == 3:
    648     difficulties = [0]
--> 650 ret = eval_class(gt_annotations, dt_annotations, current_classes, difficulties, 0,
    651                  min_overlaps, compute_aos, custom_method=custom_method)
    653 print("mAP Image BBox finished")
    654 mAP_bbox = get_m_ap(ret["precision"])

File e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\evaluate.py:44, in Evaluation.evaluate(self, result_path, current_class, score_thresh)
41 gt_annotations = kitti.get_label_annotations(self.test_annotation_file, val_image_ids)
43 evaluation_result = {}
---> 44 evaluation_result.update(get_official_eval_result(gt_annotations, dt_annotations, current_class, custom_method=0))
45 evaluation_result.update(get_official_eval_result(gt_annotations, dt_annotations, current_class, custom_method=3))
47 return evaluation_result

File e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\kitti_official_evaluate.py:744, in get_official_eval_result(gt_annotations, dt_annotations, current_classes, pr_detail_dict, custom_method)
741 compute_aos = True
742 break
--> 744 mAPbbox, mAPbev, mAP3d, mAPaos, mAPbbox_R40, mAPbev_R40, mAP3d_R40, mAPaos_R40 = do_eval(
745 gt_annotations, dt_annotations, current_classes, min_overlaps, compute_aos, pr_detail_dict=pr_detail_dict,
746 custom_method=custom_method)
748 ret_dict = {}
749 for j, curcls in enumerate(current_classes):
750 # mAP threshold array: [num_min_overlap, metric, class]
751 # mAP result: [num_class, num_diff, num_min_overlap]

File e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\kitti_official_evaluate.py:650, in do_eval(gt_annotations, dt_annotations, current_classes, min_overlaps, compute_aos, pr_detail_dict, custom_method)
647 if custom_method == 3:
648 difficulties = [0]
--> 650 ret = eval_class(gt_annotations, dt_annotations, current_classes, difficulties, 0,
651 min_overlaps, compute_aos, custom_method=custom_method)
653 print("mAP Image BBox finished")
654 mAP_bbox = get_m_ap(ret["precision"])
File e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\kitti_official_evaluate.py:536, in eval_class(gt_annotations, dt_annotations, current_classes, difficulties, metric, min_overlaps, compute_aos, num_parts, custom_method)
533 split_parts = get_split_parts(num_examples, num_parts)
535 # iou is calculated in "parts = cluster of frames"
--> 536 rets = calculate_iou_partly(dt_annotations, gt_annotations, metric, num_parts)
537 overlaps, parted_overlaps, total_dt_num, total_gt_num = rets
539 N_SAMPLE_PTS = 41 # what is this number

File e:\github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\kitti_official_evaluate.py:406, in calculate_iou_partly(gt_annotations, dt_annotations, metric, num_parts)
402 assert len(gt_annotations) == len(dt_annotations)
403 #here the erorr is occuring Valuerror: need atleast one array to stack
404 #this is commented , please uncomment after it
405 #total_dt_num = np.stack([len(a["name"]) for a in dt_annotations], 0)
--> 406 total_dt_num = np.stack([np.array(len(a["name"])) for a in dt_annotations], 0)
407 total_gt_num = np.stack([len(a["name"]) for a in gt_annotations], 0)
408 num_examples = len(gt_annotations)

File <array_function internals>:180, in stack(*args, **kwargs)

File c:\Users\Realme\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy\core\shape_base.py:422, in stack(arrays, axis, out)
420 arrays = [asanyarray(arr) for arr in arrays]
421 if not arrays:
--> 422 raise ValueError('need at least one array to stack')
424 shapes = {arr.shape for arr in arrays}
425 if len(shapes) != 1:

ValueError: need at least one array to stack

``

**Comment by andraspalffy** [comment]
Hello,

Your error stack is really hard to follow without formatting, please use either pictures or the code block on github.

However, it seems like there is a problem with your detections (dt_annotations). Are you sure they are formatted correctly? What is in that? You have to make sure to ship the detections in the correct format.

Andras

**Comment by gopalkumr** [comment]
This is where i'm getting an error.

```
from vod.evaluation import Evaluation
import os

# When the instance is created, the label locations are required.
#evaluation = Evaluation(test_annotation_file=os.path.join('example_set', 'label'))
evaluation = Evaluation(test_annotation_file=os.path.join('E:/vod/view_of_delft_PUBLIC', 'label'))

# Using the evaluate method, the model can be evaluated on the detection labels.
results = evaluation.evaluate(
    #result_path=os.path.join('example_set', 'detection'),
    result_path=os.path.join('E:/vod/view_of_delft_PUBLIC', 'detection'),
    current_class=[0, 1, 2])

print("Results: \n"
      f"Entire annotated area: \n"
      f"Car: {results['entire_area']['Car_3d_all']} \n"
      f"Pedestrian: {results['entire_area']['Pedestrian_3d_all']} \n"
      f"Cyclist: {results['entire_area']['Cyclist_3d_all']} \n"
      f"mAP: {(results['entire_area']['Car_3d_all'] + results['entire_area']['Pedestrian_3d_all'] + results['entire_area']['Cyclist_3d_all']) / 3} \n"
      f"Driving corridor area: \n"
      f"Car: {results['roi']['Car_3d_all']} \n"
      f"Pedestrian: {results['roi']['Pedestrian_3d_all']} \n"
      f"Cyclist: {results['roi']['Cyclist_3d_all']} \n"
      f"mAP: {(results['roi']['Car_3d_all'] + results['roi']['Pedestrian_3d_all'] + results['roi']['Cyclist_3d_all']) / 3} \n"
      )
```

and this evaluation function in github\view-of-delft-dataset\view-of-delft-dataset\vod\evaluation\evaluate.py

```
import os
from vod.evaluation import evaluation_common as kitti
from .kitti_official_evaluate import get_official_eval_result
from vod import get_frame_list_from_folder


class Evaluation:
    """
    Evaluation class for KITTI evaluation.
    """
    def __init__(self, test_annotation_file):
        """
        Initialize the evaluation class for KITTI evaluation.
        :param test_annotation_file: Location of the test annotation files.
        """
        self.test_annotation_file = test_annotation_file

    def evaluate(self,
                 result_path,
                 current_class=None,
                 score_thresh=-1,
                 ):
        """
        Evaluate the results.
        :param result_path: Detection labels path.
        :param current_class: Class to evaluate.
        :param score_thresh: Score threshold to use.
        :return: Results of the evaluation.
        """

        if current_class is None:
            current_class = [0, 1, 2]

        val_image_ids = get_frame_list_from_folder(result_path)

        dt_annotations = kitti.get_label_annotations(result_path, val_image_ids)

        if score_thresh > 0:
            dt_annotations = kitti.filter_annotations_low_score(dt_annotations, score_thresh)

        gt_annotations = kitti.get_label_annotations(self.test_annotation_file, val_image_ids)

        evaluation_result = {}
        evaluation_result.update(get_official_eval_result(gt_annotations, dt_annotations, current_class, custom_method=0))
        evaluation_result.update(get_official_eval_result(gt_annotations, dt_annotations, current_class, custom_method=3))

        return evaluation_result


if __name__ == '__main__':


    #evaluation = Evaluation(test_annotation_file=os.path.join('..', '..', 'example_set', 'label'))
    evaluation = Evaluation(test_annotation_file=os.path.join('..', '..', 'E:/vod/view_of_delft_PUBLIC', 'label'))

    results = evaluation.evaluate(
        #result_path=os.path.join('..', '..', 'example_set', 'detection'),
        result_path=os.path.join('..', '..', 'E:/vod/view_of_delft_PUBLIC', 'detection'),

        current_class=[0, 1, 2])

    print("Results: \n"
          f"Entire annotated area: \n"
          f"Car: {results['entire_area']['Car_3d_all']} \n"
          f"Pedestrian: {results['entire_area']['Pedestrian_3d_all']} \n"
          f"Cyclist: {results['entire_area']['Cyclist_3d_all']} \n"
          f"mAP: {(results['entire_area']['Car_3d_all'] + results['entire_area']['Pedestrian_3d_all'] + results['entire_area']['Cyclist_3d_all']) / 3} \n"
          f"Driving corridor area: \n"
          f"Car: {results['roi']['Car_3d_all']} \n"
          f"Pedestrian: {results['roi']['Pedestrian_3d_all']} \n"
          f"Cyclist: {results['roi']['Cyclist_3d_all']} \n"
          f"mAP: {(results['roi']['Car_3d_all'] + results['roi']['Pedestrian_3d_all'] + results['roi']['Cyclist_3d_all']) / 3} \n"
          )


```

And Thank you very much for such a quick response, Since I'm New in this field, Your help would grateful.

**Comment by andraspalffy** [comment]
Hi, thanks but I meant the error message, we got the code over here :D
And please print your detections before the error occurs.

**Comment by andraspalffy** [comment]
I will close this now after 3 months, please reopen it if needed.

**Comment by Petros626** [comment]
@gopalkumr You can't run this script, without preparing `gt_labels` (in KITTI camera frame format) and `detections` (the same). I'm now figuring out how to run the evaluation on my BEV labels and detections.
